File: src/main/java/com/roughlyunderscore/enchantsapi/registry/RegistrablePlaceholder.java
package com.roughlyunderscore.enchantsapi.registry;

import org.bukkit.block.Block;
import org.bukkit.entity.Entity;
import org.bukkit.entity.Player;
import org.bukkit.event.Event;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * This interface is used to register placeholders. Implement this interface to create a new placeholder and
 * override each and every method. By default, each method has a default implementation that throws an exception.
 * @see UEAPIRegistry
 * @since 2.2
 * @author Roughly_
 */
public interface RegistrablePlaceholder extends Registrable {
  /**
   * Replaces the placeholder for an event & trigger.
   * @param text the text to replace
   * @param event the event
   * @param trigger the trigger
   * @return the replaced text
   */
  default @NotNull String replaceEvent(final @NotNull String text, final @NotNull Event event, final @NotNull RegistrableTrigger trigger) {
    throw new UnsupportedOperationException("This has not been implemented.");
  }
//  /**
//   * Returns whether the placeholder is applicable for a player.
//   * If this method returns false, {@link #replaceForPlayer(Player)} should return null.
//   * @return true if the placeholder is applicable for a player, false otherwise
//   */
//  default boolean isApplicableForPlayer() {
//    throw new UnsupportedOperationException("This has not been implemented.");
//  }
//
//  /**
//   * Returns whether the placeholder is applicable for an entity.
//   * If this method returns false, {@link #replaceForEntity(Entity)} should return null.
//   * @return true if the placeholder is applicable for an entity, false otherwise
//   */
//  default boolean isApplicableForEntity() {
//    throw new UnsupportedOperationException("This has not been implemented.");
//  }
//
//  /**
//   * Returns whether the placeholder is applicable for a block.
//   * If this method returns false, {@link #replaceForBlock(Block)} should return null.
//   * @return true if the placeholder is applicable for a block, false otherwise
//   */
//  default boolean isApplicableForBlock() {
//    throw new UnsupportedOperationException("This has not been implemented.");
//  }
//
//  /**
//   * This method replaces the placeholder, having a player in mind.
//   * Check {@link #isApplicableForPlayer()} before calling this method.
//   * @param player the player
//   * @return the replaced placeholder
//   */
//  default @Nullable String replaceForPlayer(final @NotNull Player player) {
//    throw new UnsupportedOperationException("This has not been implemented.");
//  }
//
//  /**
//   * This method replaces the placeholder, having an entity in mind.
//   * Check {@link #isApplicableForEntity()} before calling this method.
//   * @param entity the entity
//   * @return the replaced placeholder
//   */
//  default @Nullable String replaceForEntity(final @NotNull Entity entity) {
//    throw new UnsupportedOperationException("This has not been implemented.");
//  }
//
//  /**
//   * This method replaces the placeholder, having a block in mind.
//   * Check {@link #isApplicableForBlock()} before calling this method.
//   * @param block the block
//   * @return the replaced placeholder
//   */
//  default @Nullable String replaceForBlock(final @NotNull Block block) {
//    throw new UnsupportedOperationException("This has not been implemented.");
//  }
}

File: src/main/kotlin/com/roughlyunderscore/registry/RegistrablePlaceholder.kt
// Copyright 2024 RoughlyUnderscore
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.roughlyunderscore.registry

import com.roughlyunderscore.annotations.Since
import com.roughlyunderscore.enums.TargetType
import org.bukkit.event.Event

/**
 * This interface is used to create placeholders.
 */
@Since("2.2")
interface RegistrablePlaceholder : Registrable {
  /**
   * Returns the string that should replace the placeholder. The implementation of this method should be done
   * so that if some failure occurs while creating a replacement for the placeholder, the method should return null.
   * For example, for a placeholder that returns the amount of money that a player has, if no economy provider is present,
   * the method will return null, otherwise it will return the stringified amount of money. [event],
   * [trigger] and [target] are used to determine the context of the placeholder, while [args] are the arguments
   * passed to the placeholder, if any.
   * @return the replacement text
   */
  @Since("2.2")
  fun replacedText(event: Event, trigger: RegistrableTrigger, target: TargetType, args: Map<String, String>): String?
}